File: src/com/suushiemaniac/cubing/bld/util/MapUtil.java
package com.suushiemaniac.cubing.bld.util;

import java.util.*;

public class MapUtil {
    public static <T extends Number> void freqMapPrint(Map<Integer, T> toPrint) {
        List<Integer> sortedKeys = new ArrayList<Integer>(toPrint.keySet());
        Collections.sort(sortedKeys, new Comparator<Integer>() {
            @Override
            public int compare(Integer o1, Integer o2) {
                return o1.compareTo(o2);
            }
        });
        for (int i : sortedKeys) System.out.println(i + ": " + toPrint.get(i));
        System.out.println("Average: " + freqMapAverage(toPrint));
    }

    public static <T extends Number> float freqMapAverage(Map<Integer, T> toAvg) {
        float sum = 0;
        int criteriaHit = 0;
        for (int i : toAvg.keySet()) {
            Number mapEntry = toAvg.get(i);
            criteriaHit += mapEntry.intValue();
            sum += i * mapEntry.floatValue();
        }
        return sum / criteriaHit;
    }
}

File: src/main/kotlin/com/suushiemaniac/cubing/bld/util/MapUtil.kt
package com.suushiemaniac.cubing.bld.util

object MapUtil {
    fun <K : Comparable<K>, V> Map<K, V>.sortedPrint() {
        this.keys.sorted().forEach {
            println("$it: ${this[it]}")
        }
    }

    fun <T : Number> Map<T, Int>.freqAverage(): Double {
        val criteriaHit = this.values.sum()
        val sum = this.entries.sumByDouble { it.key.toDouble() * it.value }

        return sum / criteriaHit
    }

    fun <K> MutableMap<K, Int>.increment(key: K) {
        this[key] = this.getOrDefault(key, 0) + 1
    }

    fun <K, V> Map<K, V?>.denullify(): Map<K, V> {
        return this.mapNotNull { (key, nullableVal) ->
            nullableVal?.let { key to it }
        }.toMap()
    }
}
